### PR TITLE
Fetcher testing

### LINKED ISSUE #133739: Repeatable fetcher tests
### Describe the bug

Fetchers are hard to test because when the output sha is verified, Nix won't download it again.
This leads to confusion and PRs that are effectively untested because the others do not realize that they need to remove the store path before testing.

Examples:

#133515

### Steps To Reproduce
1. change a fetcher
2. nix-build
3. break the fetcher
3. nix-build
4. "success" but the fetcher is broken
5. make a PR
6. wasted effort

### Expected behavior

The fetcher runs every time. This is expected but not realistic, because Nix can not guess what to re-fetch.
Instead, we can make fetcher testing easier by adding fetcher tests following the pattern:

```nix
let
  plain = fetchFoo { ... };

  computeSaltedHash = salt: runCommand "source" {} ''
    cp -r --no-preserve-mode ${plain} src
    echo ${salt} >src/fetcher-test-salt

    # not sure which command this should be
    compute-nar-hash src >$out
  '';

  fetchWithSaltedHash = salt: plain.override {
    postFetch = "echo ${salt} >$out/fetcher-test-salt";
    sha256 = builtins.readFile computeSaltedHash; # "import from derivation"
  }

in
f {
  tests.fetch-salted = fetchWithSaltedHash;
}
```

So this uses IFD which is probably ok for ofborg?
Regarding `f`; instead of `passthru` somewhere, we'll have to use `__functor` to set `tests.fetch-salted` on the fetchers because they're functions.
The salt may need to be passed externally in case of pure evaluation. We could make it a Nixpkgs `config` parameter that defaults to `builtins.currentTime`.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
 - fetchgit
 - fetchFromGitHub
 - fetchsvn
 - dockerTools.pullImage
 - fetch*
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation for this change

Make maintenance easier on the fetchers and avoid problems.

The `packageTests` attribute will serve two purposes:
 - for ofborg: allow it to build tests in a __functor attrset like `fetchgit`
 - for hydra, potentially, build the package tests. Enabling these is not part of this PR.

Closes #133739

Refs #128749 #135881 or really all of [topic: fetch](https://github.com/NixOS/nixpkgs/pulls?q=is%3Apr+is%3Aopen++label%3A%226.topic%3A+fetch%22+)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
Not picked up automatically by ofborg, as expected. So,

@GrahamcOfBorg build packageTests.fetchgit

**Comment by Artturin** [comment]
probably needs documentation in `doc/builders/fetchers.chapter.md`?

**Comment by Artturin** [comment]
can't get this to work

```diff
diff --git a/pkgs/build-support/fetchfirefoxaddon/tests.nix b/pkgs/build-support/fetchfirefoxaddon/tests.nix
new file mode 100644
index 00000000000..05ee43aba29
--- /dev/null
+++ b/pkgs/build-support/fetchfirefoxaddon/tests.nix
@@ -0,0 +1,10 @@
+{ invalidateFetcherByDrvHash, fetchFirefoxAddon, ... }:
+
+{
+  simple = invalidateFetcherByDrvHash fetchFirefoxAddon {
+    name = "image-search-options";
+    # Chosen because its only 147KB
+    url = "https://addons.mozilla.org/firefox/downloads/file/3059971/image_search_options-3.0.12-fx.xpi";
+    sha256 = "1h768ljlh3pi23l27qp961v1hd0nbj2vasgy11bmcrlqp40zgvnr";
+  };
+}
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 161c48a3d48..d14a587eb11 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -516,7 +516,10 @@ with pkgs;
 
   fetchhg = callPackage ../build-support/fetchhg { };
 
-  fetchFirefoxAddon = callPackage ../build-support/fetchfirefoxaddon {};
+  fetchFirefoxAddon = callPackage ../build-support/fetchfirefoxaddon { }
+    // {
+      tests = callPackages ../build-support/fetchfirefoxaddon/tests.nix;
+  };
 
   # `fetchurl' downloads a file from the network.
   fetchurl = if stdenv.buildPlatform != stdenv.hostPlatform
```

```sh
nix-repl> packageTests.fetchgit
{ recurseForDerivations = true; simple = «derivation /nix/store/5rl9by8nifg2k25rzp3yw5fdj303r6qn-nix-source-salted-v839ywjgprrp.drv»; }

nix-repl> packageTests.fetchFirefoxAddon
error: value is a function while a set was expected
```

**Comment by roberth** [comment]
Diffception.

> ```diff
> -+      tests = callPackages ../build-support/fetchfirefoxaddon/tests.nix;
> ++      tests = callPackages ../build-support/fetchfirefoxaddon/tests.nix { };
> ```

I also fall into this trap every now and then, fwiw.

> documentation

yes.

File: doc/builders/fetchers.chapter.md [hunk 1]
```diff
@@ -20,7 +28,7 @@ The main difference between `fetchurl` and `fetchzip` is in how they store the c
 
 `fetchpatch` works very similarly to `fetchurl` with the same arguments expected. It expects patch files as a source and performs normalization on them before computing the checksum. For example it will remove comments or other unstable parts that are sometimes added by version control systems and can change over time.
 
-Other fetcher functions allow you to add source code directly from a VCS such as subversion or git. These are mostly straightforward nambes based on the name of the command used with the VCS system. Because they give you a working repository, they act most like `fetchzip`.
```
Comment (roberth) [inline doc/builders/fetchers.chapter.md:23]: This was 90% stating the obvious.

File: doc/builders/fetchers.chapter.md [hunk 2]
```diff
@@ -1,8 +1,16 @@
 # Fetchers {#chap-pkgs-fetchers}
 
-When using Nix, you will frequently need to download source code and other files from the internet. Nixpkgs comes with a few helper functions that allow you to fetch fixed-output derivations in a structured way.
+When using Nix, you will frequently need to download source code and other files from the internet. For this purpose, Nix provides the [_fixed output derivation_](https://nixos.org/manual/nix/stable/#fixed-output-drvs) feature and Nixpkgs provides various functions that implement the actual fetching from various protocols and services.
 
-The two fetcher primitives are `fetchurl` and `fetchzip`. Both of these have two required arguments, a URL and a hash. The hash is typically `sha256`, although many more hash algorithms are supported. Nixpkgs contributors are currently recommended to use `sha256`. This hash will be used by Nix to identify your source. A typical usage of fetchurl is provided below.
```
Comment (roberth) [inline doc/builders/fetchers.chapter.md:5]: > The two fetcher primitives

There's no such thing.

**Review by roberth — COMMENTED** [review]
I've cleaned up some minor docs issues while I was at it. I hope those will not hinder this pr.

**Comment by Artturin** [comment]
will having all the expressions in `packageTests.X` slow something(eval?) down
would it be possible to only have the ones that have tests be there

~is `packageTests` even needed since 
`nix build ".#fetchgit.tests.simple"` works?~

stable nix doesn't work like that`nix-build -A fetchgit.tests.simple`

**Comment by roberth** [comment]
> will having all the expressions in `packageTests.X` slow something(eval?) down

This adds about as much overhead as a package would. In most use cases it would go unnoticed like any other package. In cases where the whole of Nixpkgs is traversed, those traversals look for the `recurseForDerivations` attribute, which is unset inside `packageTests`, causing it to be skipped. This works for `nix-build` (which shouldn't be used on all of Nixpkgs anyway) and it works for Hydra/`release.nix`. In this case, the attribute _names_ are copied into memory once more, but this is still insignificant and the same happens for `pkgsStatic` etc.

**Review by Artturin — APPROVED** [review]

**Comment by Artturin** [comment]
@ofborg eval

**Comment by Artturin** [comment]
How would one run all the packageTests?

**Comment by roberth** [comment]
@Artturin you generally would, if you mean literally all.

@GrahamcOfBorg build packageTests.fetchgit

**Comment by roberth** [comment]
Not sure if I tested this or just assumed it still uses `nix-build` _and_ suffers from its functor problem.

@GrahamcOfBorg build fetchgit.tests

**Comment by roberth** [comment]
Confirmed. Ofborg tries to autocall the functor instead of using the attribute.

```
Cannot nix-instantiate `fetchgit.tests' because:
error: cannot auto-call a function that has an argument without a default value ('url')
```

I'll leave ofborg out of scope for now and merge this when ofborg is done.

**Pull request merged**